### PR TITLE
Triton deployment improvements for in-framework models

### PR DESCRIPTION
**Pull request by jukim-nv**
# What does this PR do ?

Adding better support for triton deployment of in-framework models using `MegatronLLMDeployable`.

**Collection**: [Note which collection this PR will affect]
scripts/deploy
tests/deploy

# Changelog 
- Add `NemoQueryLLMPyTorch` class for easy query of `MegatronLLMDeployable`
- Changes to `nemo_export.py` to accurately test `MegatronLLMDeployable` models
- A separate `deploy_inframework_triton.py` deployment script to easily deploy `MegatronLLMDeployable` on a Triton server

# Usage

```
python scripts/deploy/nlp/deploy_inframework_triton.py -nc <path to .nemo checkpoint> -tmn TritonModelName 

python tests/export/nemo_export.py --checkpoint_dir <path to .nemo checkpoint> --model_type gpt --min_tps 1 --max_output_len 24 --test_deployment False --in_framework True --model_name TritonModelName
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [x] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Review by oyilmaz-nvidia — COMMENTED** [review]
Ran this python tests/export/nemo_export.py --model_name LLAMA2-7B-base --existing_test_models --min_tps 1 --in_framework True and got the following error

[rank0]: Traceback (most recent call last):
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/pytriton/client/client.py", line 661, in _infer
[rank0]:     response = self._infer_client.infer(
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/tritonclient/http/_client.py", line 1482, in infer
[rank0]:     _raise_if_error(response)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/tritonclient/http/_utils.py", line 69, in _raise_if_error
[rank0]:     raise error
[rank0]: tritonclient.utils.InferenceServerException: [400] [request id: 0] unexpected shape for input 'max_length' for model 'LLAMA2-7B-base'. Expected [-1,1], got [1]. NOTE: Setting a non-zero max_batch_size in the model config requires a batch dimension to be prepended to each input shape. If you want to specify the full shape including the batch dim in your input dims config, try setting max_batch_size to zero. See the model configuration docs for more info on max_batch_size.

[rank0]: The above exception was the direct cause of the following exception:

[rank0]: Traceback (most recent call last):
[rank0]:   File "/opt/NeMo/tests/export/nemo_export.py", line 909, in <module>
[rank0]:     run_inference_tests(args)
[rank0]:   File "/opt/NeMo/tests/export/nemo_export.py", line 777, in run_inference_tests
[rank0]:     result_dic[tps] = run_existing_checkpoints(
[rank0]:   File "/opt/NeMo/tests/export/nemo_export.py", line 482, in run_existing_checkpoints
[rank0]:     return run_in_framework_inference(
[rank0]:   File "/opt/NeMo/tests/export/nemo_export.py", line 561, in run_in_framework_inference
[rank0]:     output_deployed = nq.query_llm(
[rank0]:   File "/opt/NeMo/nemo/deploy/nlp/query_llm.py", line 123, in query_llm
[rank0]:     result_dict = client.infer_batch(**inputs)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/pytriton/client/client.py", line 603, in infer_batch
[rank0]:     return self._infer(inputs or named_inputs, parameters, headers)
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/pytriton/client/client.py", line 678, in _infer
[rank0]:     raise PyTritonClientInferenceServerError(
[rank0]: pytriton.client.exceptions.PyTritonClientInferenceServerError: Error occurred during inference request. Message: [request id: 0] unexpected shape for input 'max_length' for model 'LLAMA2-7B-base'. Expected [-1,1], got [1]. NOTE: Setting a non-zero max_batch_size in the model config requires a batch dimension to be prepended to each input shape. If you want to specify the full shape including the batch dim in your input dims config, try setting max_batch_size to zero. See the model configuration docs for more info on max_batch_size.
Signal (2) received.

But this worked

python tests/export/nemo_export.py --model_name llama --model_type llama --min_tps 1 --checkpoint_dir /opt/checkpoints/LLAMA2-7B-base/LLAMA2-7B-base-1.nemo --in_framework True

File: tests/export/nemo_export.py [hunk 1]
```diff
@@ -101,52 +101,82 @@
...
+                    deployed_output = deployed_output["sentences"][0][0][len(prompt) :].decode().strip().lower()
+                else:
+                    deployed_output = nq.query_llm(
+                        prompts=[prompt],
+                        max_output_len=1,
+                        top_k=1,
+                        top_p=0,
+                        temperature=0.1,
+                        task_id=task_ids,
+                    )
```
[the remark this answers is not in the document]
Comment (jukim-nv) [reply tests/export/nemo_export.py:168]: This should be ok, `nq` in this case is a `NemoQueryLLM`, not `NemoQueryLLMPyTorch`, and so this argument list is correct.

File: tests/export/nemo_export.py [hunk 2]
```diff
@@ -534,14 +564,15 @@
...
-            prompts=[prompts],
-            top_k=top_k,
-            top_p=top_p,
-            temperature=temperature,
+            prompts=prompts, top_k=top_k, top_p=top_p, temperature=temperature, max_length=max_output_len
         )
+        output_deployed = output_deployed["sentences"]
+        # MegatronLLMDeployable will return the prompt + generated output, so cut off the prompt
+        for i, output in enumerate(output_deployed):
+            output = output[len(prompts[i]) :]
```
[the remark this answers is not in the document]
Comment (jukim-nv) [reply tests/export/nemo_export.py:575]: This code should still modify `output_deployed` so I think it is ok.

**Comment by oyilmaz-nvidia** [comment]
Could you please address all of these CodeQL reviews?

**Review by oyilmaz-nvidia — APPROVED** [review]

**Pull request merged**